include/kllvm/codegen/Decision.h

```diff
@@ -69,60 +82,39 @@ class SwitchNode : public DecisionNode {
   const std::vector<DecisionCase> &getCases() const { return cases; }
   
   virtual void codegen(Decision *d, llvm::StringMap<llvm::Value *> substitution);
-};
-
-class MakeLiteralNode : public DecisionNode {
-private:
-  std::string name;
-  ValueType cat;
-  std::string literal;
-  DecisionNode *child;
-
-  MakeLiteralNode(
-    const std::string &name,
-    ValueType cat,
-    const std::string literal,
-    DecisionNode *child) :
-      name(name),
-      cat(cat),
-      literal(literal),
-      child(child) {}
-
-public:
-  static MakeLiteralNode *Create(
-      const std::string &name,
-      ValueType cat,
-      const std::string literal,
-      DecisionNode *child) {
-    return new MakeLiteralNode(name, cat, literal, child);
-  }
-
-  virtual void codegen(Decision *d, llvm::StringMap<llvm::Value *> substitution);
+  virtual void collectUses(std::set<std::string> &vars) { if(cases.size() != 1 || cases[0].getConstructor()) vars.insert(name); for (auto _case : cases) { _case.getChild()->collectUses(vars); } }
+  virtual void collectDefs(std::set<std::string> &vars) { for (auto _case : cases) { vars.insert(_case.getBindings().begin(), _case.getBindings().end()); _case.getChild()->collectDefs(vars); } }
 };
 
 class MakePatternNode : public DecisionNode {
 private:
   std::string name;
   KOREObjectPattern *pattern;
+  std::vector<std::string> uses;
   DecisionNode *child;
 
   MakePatternNode(
     const std::string &name,
     KOREObjectPattern *pattern,
+    std::vector<std::string> &uses,
     DecisionNode *child) :
       name(name),
       pattern(pattern),
+      uses(uses),
       child(child) {}
 
 public:
   static MakePatternNode *Create(
       const std::string &name,
       KOREObjectPattern *pattern,
+      std::vector<std::string> &uses,
       DecisionNode *child) {
-    return new MakePatternNode(name, pattern, child);
+    return new MakePatternNode(name, pattern, uses, child);
   }
 
   virtual void codegen(Decision *d, llvm::StringMap<llvm::Value *> substitution);
+  virtual void collectUses(std::set<std::string> &vars) { vars.insert(uses.begin(), uses.end()); child->collectUses(vars); }
+  virtual void collectDefs(std::set<std::string> &vars) { vars.insert(name); child->collectDefs(vars); }
```

Comment:
Is `collectDef` supposed to collect everything bound by this node? I would think the `name` being matched already existed above, while the variables contained in `pattern` would be the ones that are bound.

Reply:
collectDef collects every variable that is introduced recursively by this node. MakePattern creates a new pattern, so name is a def, not a use.

Reply:
Ah, I was confusing it with matching on a pattern.